### ISSUE TITLE
Wrong IP when connecting to a server with the same public ip address

### ISSUE THREAD

**Issue by ghost** [VeniceUnleashed/VeniceUnleashed#566]
When connecting to a server with the same public IP-address as the client(NAT), the client connects to 127.0.0.1 instead of the right local one.

I am using NordVPN to get out of my local network and the bug is gone. (TEMP FIX!)

**Comment by ghost** [comment]
![image](https://user-images.githubusercontent.com/943976/102938494-b8c64f80-44ac-11eb-81f3-db31267888bc.png)

**Comment by OrfeasZ** [comment]
This is a known issue if your server or client is running under a VM. If it is, try changing the mac address of the VM's network adapter and trying again.

**Comment by ghost** [comment]
The server is running in a vm, but not on the same pc as the client. Are you shure this would help? It’s running on a proxmox vhost, not something like virtualbox that does nat „things“.

**Comment by OrfeasZ** [comment]
Yep, just try changing the mac address of the network adapter from proxmox to something random and try again.

**Comment by ghost** [comment]
I changed the mac address to a random one and nothing changed(as expected).
The VM is in another subnet as the client, but still NATed to the same external ip address.
They can not communicate directly via MAC-address on layer 2.

**Comment by OrfeasZ** [comment]
VU excludes certain VM network adapters from it's NAT detection based on their MAC address prefix, which is why I asked you to change it. Did you maintain the 3-byte prefix or did you make the entire address random?

**Comment by ghost** [comment]
> VU excludes certain VM network adapters from it's NAT detection based on their MAC address prefix, which is why I asked you to change it. Did you maintain the 3-byte prefix or did you make the entire address random?
> […](#)
> On Dec 23, 2020, 16:17, at 16:17, timniklas ***@***.***> wrote: I changed the mac address to a random one and nothing changed(as expected). The VM is in another subnet as the client, but still NATed to the same external ip address. They can not communicate directly via MAC-address on layer 2. -- You are receiving this because you commented. Reply to this email directly or view it on GitHub: [#566 (comment)](https://github.com/EmulatorNexus/VeniceUnleashed/issues/566#issuecomment-750319523)

changed it to a completly random one. b2:dc:97:29:04:04

**Comment by OrfeasZ** [comment]
Alright that sounds like an issue then. I'll investigate when I have some time.

**Comment by nik-johnson-net** [comment]
Hey @OrfeasZ, looks like I'm having a similar issue.

Can you explain further which MAC addresses are excluded from NAT detection, and how that might affect local players?

My network setup relevant to this:
One public IP address, NAT'd to multiple subnets. One subnet includes my desktop, another includes my server. With NAT loopback enabled, this setup works fine for other servers. I can access them via the public IP address, and so can others outside my NAT.

For VU, others can access the server I'm hosting, but my local client attempts to connect to `127.0.0.1`, which is clearly incorrect as the server exists on a different computer. It's MAC address is ` 02:00:00:0c:43:5d`. I can connect to the RCON console from my desktop just fine.

A nice workaround would be to manually join a server by IP Address and save to a "favorites" list, but my assumption is there's something in the client altering the master list IP address to 127.0.0.1 incorrectly.

Happy to give more information or chat on discord to debug and solve this issue.

**Comment by Bonn93** [comment]
I have the same issue, its feels like there's some split brain/async routing going on. I've been using PFsense with multiple VLANS, as this is a windows service, this bounces via my DMZ. Firewall rules/NAT etc are all configured. External players to my network connect fine, its only internal. 

Other game services like Topaz for FFXI offer a "hairpin NAT" option which helps navigate these issues.

**Comment by Turranx** [comment]
I am also experiencing this problem, and like everyone else, my VU server and workstation share the same public IP, but are located in different internal subnets.  What follows is a description of my environment.  All IP addresses have been changed to protect the guilty.

Network Structure:
- Router: OPNsense
  - Internet gateway (NAT)
  - Routing and firewall services between internal subnets
- DMZ subnet: 192.168.50.0/24
- Workstation subnet: 192.168.30.0/24

Venice Unleashed Server:
- Ubuntu 20.04 LTS
- 192.168.50.5
  - Ports 7948 and 25200 on public IP are brought to the server via NAT Port Forward
  - [NAT "reflection" enabled](https://forum.opnsense.org/index.php?topic=1884.msg5861#msg5861)
- 52:5f:61:30:9b:4a

Workstation:
- Windows 10
- 192.168.30.8

NOTE:  NAT "reflection" is supposed to allow my workstations sitting on 192.168.30.0/24 to access my servers in 192.168.50.0/24 by using the public IP address.

For a little more background (with pictures!), see my [post on the Venice Unleashed Community forum](https://community.veniceunleashed.net/t/server-shows-in-list-but-connections-always-fail/1321).

I've tried various combinations of command line arguments to see what might work.  When used, every item here would have started with "wine ~/vu/client/vu.com -gamepath ~/bf3 -serverInstancePath ~/vu/instance -server -dedicated"
- -headless -listen 0.0.0.0:25200 -mHarmonyPort 7948
- -headless -listen 0.0.0.0:25200 -mHarmonyPort 7948 -joinaddr 192.168.50.5
- -headless -listen 0.0.0.0:25200 -mHarmonyPort 7948 -joinaddr <Public_IP>
- -headless -listen 192.168.50.5:25200 -mHarmonyPort 7948 -joinaddr "192.168.50.5"
- -headless -listen 192.168.50.5:25200 -mHarmonyPort 7948 -joinaddr 192.168.50.5
- -headless -listen 192.168.50.5:25200 -mHarmonyPort 7948 -joinaddr '<Public_IP>'
- -headless -listen 192.168.50.5:25200 -mHarmonyPort 7948 -joinaddr <Public_IP>
- -headless -listen <Public_IP>:25200 -mHarmonyPort 7948 -joinaddr <Public_IP>
- -listen 0.0.0.0:25200 -mHarmonyPort 7948
- -listen 192.168.50.5:25200 -mHarmonyPort 7948

**Comment by OrfeasZ** [comment]
I've identified the issue and it will be resolved in an upcoming build (probably live sometime next week). Keep in mind that you'll need to have hairpinning properly set up in order to be able to connect to your server.

**Comment by Bonn93** [comment]
@OrfeasZ can you share any detail or config needed for the fix too?

**Comment by charlieparkes** [comment]
In my case, both the server and client are in the same subnet, but the server service itself is running in docker (in an overlay network). Might this be the same issue?

**Comment by zapotah** [comment]
This is not an "issue" per say. Its how service discovery works with things like these. (i admittedly didnt bother checking if the on-link discovery is SSDP or what so it could be routed but thats quite a bit more advanced than what i propose.) @OrfeasZ the sane option for those who know what theyre doing would be to have an option to register a dns record in zeus for service resolution instead of blindly taking the address from the server socket to zeus. This would allow setting up a split scope for internal connectivity. NAT hairpinning is always an ugly hack and introduces other problems.

**Comment by muppet99** [comment]
Got your server nicely up and running. Kudo's for the nice work. You got some attention on a dutch tech website: https://tweakers.net/nieuws/176222/battlefield-3-modplatform-maakt-gebruik-van-eigen-servers-en-mods-mogelijk.html

But also trying to get around this bug. Can't join my own server. As soon I use a VPN the problem is over. If you need any extra information besides what is already mentioned, please let me know.

**Comment by mvivaldi** [comment]
> 
> 
> This is not an "issue" per say. Its how service discovery works with things like these. (i admittedly didnt bother checking if the on-link discovery is SSDP or what so it could be routed but thats quite a bit more advanced than what i propose.) @OrfeasZ the sane option for those who know what theyre doing would be to have an option to register a dns record in zeus for service resolution instead of blindly taking the address from the server socket to zeus. This would allow setting up a split scope for internal connectivity. NAT hairpinning is always an ugly hack and introduces other problems.

I have the same problem, and I think the dns solution is a good one.

P.S thank you all for the hard work.. the possibility to host my BF3 server is something unbelievable!

**Comment by OrfeasZ** [comment]
This has been fixed in build `17485` and should work as expected for people who have hairpin NAT set up correctly.

For everyone else, we have introduced a new server launch argument: [-joinhost <hostname>](https://docs.veniceunleashed.net/general/args/#server-arguments).

When you specify this argument when launching your server any clients attempting to connect will skip any NAT detection and attempt to connect directly to the hostname you've provided.
